Compute broadcast-join segmentMapFn only once per worker.

gianm opened the pull request:
This patch introduces "processor managers" to processor factories, as a replacement for the sequence of processors. Processor managers can use the results of earlier processors to influence the creation of later processors, which provides us with the building block we need to ensure that broadcast join data is only read once.

In particular, when broadcast join is happening, the `BaseFrameProcessorFactory` now uses a `ChainedProcessorManager` to first run `BroadcastJoinSegmentMapFnProcessor` (in a single thread), and *then* run all of the regular processors (possibly multithreaded).

LakshSingla commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerMemoryParameters.java:
```diff
@@ -130,11 +131,11 @@ public class WorkerMemoryParameters
+   * a 2x expansion, and object and index overhead must be considered on top of that. So we use a value somewhat
+   * lower than 0.5.
    */
   static final double BROADCAST_JOIN_MEMORY_FRACTION = 0.3;
```
Passing comment: Since the broadcast table is computed once, perhaps we can rework the space allocated for the broadcast to be a fraction of the whole than per bundle?

gianm replied:
This makes sense in principle, although I want to hold off on it, because right now the worker bundle is mostly taken up by items associated with the shuffle. Moving this to the worker bundle would crowd that out.

I have been thinking generally of increasing the size of the worker bundle and decreasing the size of the processing bundle, because the processors get to use processing buffers, which are pretty big. There isn't much need for much extra memory, other than `segmentGenerator`. But we can have that use worker-bundle memory by limiting the # of segments build concurrently per task to 1. Which is effectively happening anyway for tasks.

Anyway, point is, I agree with you but I think it's not quite the right time 🙂

adarshsanjeev commented:
I'm still reviewing some of the classes, but this LGTM overall!

cryptoe commented on processing/src/main/java/org/apache/druid/frame/processor/manager/ProcessorManager.java:
```diff
@@ -0,0 +1,76 @@
+
+  default <R2> ProcessorManager<T, R2> withAccumulation(
+      R2 initialResult,
+      BiFunction<R2, T, R2> accumulateFn
```
What does T represent here. 
R seems to be the final result type. 

Is r2 the intermediate result of processor's ?

Could you please add java doc to this method. It seems like this would be called once to chain process managers.

gianm replied:
I added some javadoc to clarify what these mean.

cryptoe commented on processing/src/main/java/org/apache/druid/frame/processor/manager/SequenceProcessorManager.java:
```diff
@@ -0,0 +1,103 @@
+      CloseableUtils.closeAndWrapExceptions(yielder);
+      done = true;
+      yielder = null;
+    }
```
In this close call, it might me that the yielder is not initialized, and then the done variable will not be set. 
It will result in this erroneous flow to pass. 
```SPM spm=new SPM();
spm.close();
spm.next();
```
here spm.next() should throw a ` NoSuchElementException();`

I think done=true should be moved out of this null check.

cryptoe commented on processing/src/main/java/org/apache/druid/frame/processor/manager/ProcessorManager.java:
```diff
@@ -0,0 +1,76 @@
+  /**
+   * Called after all procesors are done, prior to {@link #close()}, to retrieve the result of this computation.
+   */
+  R result();
```
Is this a contract that results should be called only after next returns and empty optional. 

If yes then we need to update sequenceProcessorManager#result method with a done check.

gianm replied:
Yes, although behavior is undefined if the contract is not adhered to, so implementations of ProcessorManager are not required to verify that the methods are being called properly.

Reason for this is it could muck up the logic in the managers, making them harder to understand, and there's only one caller anyway (`RunAllFullyWidget`). Which we do have unit tests for, to ensure it behaves properly.

cryptoe commented on processing/src/main/java/org/apache/druid/frame/processor/manager/AccumulatingProcessorManager.java:
```diff
@@ -0,0 +1,78 @@
+  @Override
+  public ListenableFuture<Optional<ProcessorAndCallback<T>>> next()
+  {
+    return FutureUtils.transform(
```
the next should have a currrentResult!=null check since the caller can call a close after the initialization.

gianm replied:
Added a check for whether the processor is closed, although it isn't strictly necessary, since behavior is undefined in that case. But it doesn't muck up the logic too much so I figured it's ok.

cryptoe commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java:
```diff
@@ -0,0 +1,104 @@
+
+  @Override
+  public void close()
+  {
```
I guess we are missing the close case here as well.

gianm replied:
What do you mean by this?

cryptoe replied:
```
CPM cpm=new CPM();
cpm.close();
cpm.next();
```

gianm replied:
Ah, got it. It's not required that the processor handle this case, because behavior of `next` is undefined after `close` is called. Nevertheless, in this scenario it isn't too bad to handle it, so I added some defensive handling. Thanks for clarifying.

cryptoe reviewed (commented):
Cool PR. Left some comments.

adarshsanjeev commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerImpl.java:
```diff
@@ -1152,7 +1152,7 @@ private void makeShuffleOutputChannelFactory(boolean isFinalStage)
     }
 
-    private <FactoryType extends FrameProcessorFactory<I, WorkerClass, T, R>, I, WorkerClass extends FrameProcessor<T>, T, R> void makeAndRunWorkProcessors()
+    private <FactoryType extends FrameProcessorFactory<T, R, I>, T, R, I> void makeAndRunWorkProcessors()
```
Nit: this might be easier to read with more detailed generic names

gianm replied:
I updated the names to be clearer and added javadoc.

gianm commented on processing/src/main/java/org/apache/druid/frame/processor/manager/SequenceProcessorManager.java:
```diff
@@ -0,0 +1,103 @@
+      }
+      catch (Throwable e) {
+        // Some problem with yielder.get() or yielder.next(null). Close the yielder and mark us as done.
+        done = true;
+        CloseableUtils.closeAndSuppressExceptions(yielder, e::addSuppressed);
+        yielder = null;
```
It's a little different, this one adds a suppressed exception to `e` whereas `close()` would throw it.

gianm commented on processing/src/main/java/org/apache/druid/frame/processor/manager/SequenceProcessorManager.java:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  @Override
+  public Long result()
```
I updated the docs of `ProcessorManager#result` to call out that behavior is undefined in this case. I don't think it's necessary for the managers to all have code that checks that callers are doing their stuff properly. There's only one caller (`RunAllFullyWidget`) and we have unit tests for that.

gianm commented on processing/src/main/java/org/apache/druid/frame/processor/manager/ProcessorManager.java:
```diff
@@ -0,0 +1,76 @@
+ * Used by {@link FrameProcessorExecutor#runAllFully} to manage the launching of processors. Processors returned by
+ * this class may run concurrently with each other.
+ *
+ * This interface allows for simple sequences of processors, such as {@link ProcessorManagers#of(Sequence)}. It also
+ * allows for situations where later processors depend on the results of earlier processors. (The result of earlier
+ * processors are made available to the manager through {@link ProcessorAndCallback#onComplete(Object)}.)
```
I added a little text about this. The idea is that `next()` returns a future, and so the future resolution can be delayed until the "earlier" processors run.

gianm commented on processing/src/main/java/org/apache/druid/frame/processor/manager/ProcessorManager.java:
```diff
@@ -0,0 +1,76 @@
+  ListenableFuture<Optional<ProcessorAndCallback<T>>> next();
+
+  /**
+   * Called after all procesors are done, prior to {@link #close()}, to retrieve the result of this computation.
```
I updated the docs of `ProcessorManager#result` to call out that behavior is undefined in this case. I don't think it's necessary for the managers to all have code that checks that callers are doing their stuff properly. There's only one caller (`RunAllFullyWidget`) and we have unit tests for that.

gianm commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerImpl.java:
```diff
@@ -1152,7 +1152,7 @@ private void makeShuffleOutputChannelFactory(boolean isFinalStage)
     }
 
-    private <FactoryType extends FrameProcessorFactory<I, WorkerClass, T, R>, I, WorkerClass extends FrameProcessor<T>, T, R> void makeAndRunWorkProcessors()
+    private <FactoryType extends FrameProcessorFactory<T, R, I>, T, R, I> void makeAndRunWorkProcessors()
```
I expanded these to more clear names and added some javadoc.

gianm commented:
Pushed up a patch that addresses all comments so far, except this one https://github.com/apache/druid/pull/15007#discussion_r1341255127, because I didn't quite understand it.

gianm commented:
Pushed up another patch addressing https://github.com/apache/druid/pull/15007#discussion_r1341255127.

gianm commented on processing/src/main/java/org/apache/druid/frame/processor/manager/SequenceProcessorManager.java:
```diff
@@ -0,0 +1,110 @@
+        // Not closed yet, but done.
+        throw new NoSuchElementException();
+      }
+    } else if (yielder.isDone()) {
```
False alarm, I believe, since `done` is `false` in this path so `yielder` must be nonnull.

cryptoe reviewed (approved):
Thanks for the patch @gianm. LGTM!!
